File: 백준/Silver/2178. 미로 탐색/미로 탐색.py
import sys
input=sys.stdin.readline

def function():
    # write down code
    # 1. n, m 행 열
    # 2. 1: 이동 가능, 0: 이동 불가능
    # 3. 1, 1 -> n,m으로 이동할떄 지나하야하는 최소 칸 수 return
    # 특이사항: 항상 도착위치로 이동 가능함
    # 최소 칸으로 이동해야 하기에 왔던 칸으로 돌아갈 이유 없음

    n, m = map(int, input().split())

    graph = [input().strip() for _ in range(n)]
    visitied = [[False]*m for _ in range(n)]

    return bfs(graph, visitied)

def bfs(graph, visited):
    from collections import deque
    # 상, 하, 좌, 우
    dy = [-1, 1, 0, 0]
    dx = [0, 0, 1, -1]

    n, m = len(graph), len(graph[0])

    q = deque([[0, 0, 1]])
    visited[0][0] = True

    while q:
        y, x, cnt = q.popleft()

        if y == n - 1 and x == m - 1:
            return cnt
        for i in range(4): # 1. 현재 노드에서 상, 하, 좌, 우 탐색
            ny, nx = y+dy[i], x+dx[i]
            if 0 <= ny < n and 0 <= nx < m and graph[ny][nx] == '1' and not visited[ny][nx]: # 범위를 벗어나지 않는지 and 방문 가능한지 판단
                q.append([ny, nx, cnt+1])
                visited[ny][nx] = True



if __name__ == "__main__":
    print(function())

File: 백준/Silver/2178. 미로 탐색/미로 탐색.java
import java.util.*;
import java.io.*;

public class Main {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;

    static int[] dy = {1,-1,0,0};
    static int[] dx = {0,0,1,-1};

    public static class Point {
        int y;
        int x;
        int cnt = 0;
        Point(int a, int b, int c){
            y=a;
            x=b;
            cnt=c;
        }
    }

    public static int bfs(Point start){
        Deque<Point> q = new LinkedList<>();
        q.addFirst(start);
        visited[start.y][start.x] = true;

        int res=0;

        while(!q.isEmpty()){
            Point node = q.pollLast();
            
            if(node.x==n-1 && node.y==m-1){
                res = node.cnt;
                break;
            }

            for(int i=0; i<4; i++){
                int ny = node.y + dy[i];
                int nx = node.x + dx[i];
                
                if(0 <= nx && nx < n && 0 <= ny && ny < m && graph[ny][nx]=='1' && !visited[ny][nx]){
                    q.addFirst(new Point(ny, nx, node.cnt+1));
                    visited[ny][nx] = true;
                }
            }
        }
        
        return res;
    }

    static int m, n;
    static char[][] graph;
    static boolean[][] visited;
    public static void main(String[] args) throws IOException{
        st = new StringTokenizer(br.readLine());
        m = Integer.parseInt(st.nextToken());
        n = Integer.parseInt(st.nextToken());
        graph = new char[m][n];
        visited = new boolean[m][n];

        for(int i=0; i<m; i++) graph[i] = br.readLine().toCharArray();

        int res = bfs(new Point(0, 0, 1));

        System.out.print(res);
        
    }
}